### PR TITLE
[config] configuration yaml generation

### PR DESCRIPTION
**Pull request by Peguen**
### Description
Use defines to generate a yaml stringstream

### CODE REVIEW

File: ecal/core/include/ecal/config/configuration.h [hunk 1]
```diff
@@ -48,31 +48,30 @@ namespace eCAL
...
-    Cli::Configuration               command_line_arguments{};
+    TransportLayer::Configuration transport_layer{};
+    Registration::Configuration   registration{};
+    Monitoring::Configuration     monitoring{};
+    Subscriber::Configuration     subscriber{};
+    Publisher::Configuration      publisher{};
+    Time::Configuration           timesync{};
+    Service::Configuration        service{};
+    Application::Configuration    application{};
+    Logging::Configuration        logging{};
```
Comment (KerstinKeller) [inline ecal/core/include/ecal/config/configuration.h:59]: You don't need to initialize objects explicitly, the will be default initialized. I guess it's only necessary when they need a non default constructor. However, it's also not "wrong" to do it.

File: ecal/core/include/ecal/config/registration.h [hunk 2]
```diff
@@ -39,39 +39,39 @@ namespace eCAL
...
+        UDP::Configuration udp;                  /*!< UDP based registration configuration */
       };
     }
 
     struct Configuration
     {
-      unsigned int         registration_timeout;    //!< Timeout for topic registration in ms (internal) (Default: 60000)
-      unsigned int         registration_refresh;    //!< Topic registration refresh cylce (has to be smaller then registration timeout!) (Default: 1000)                                   
+      unsigned int         registration_timeout { 60000U }; //!< Timeout for topic registration in ms (internal) (Default: 60000)
+      unsigned int         registration_refresh { 1000U };  //!< Topic registration refresh cylce (has to be smaller then registration timeout!) (Default: 1000)
```
Comment (KerstinKeller) [inline ecal/core/include/ecal/config/registration.h:67]: In a following PR we could change timeouts to std::Chrono objects? At least it might make it easier for the user if he sets the values programatically...

File: ecal/core/src/config/default_config.h [hunk 3]
```diff
@@ -42,12 +42,12 @@ const std::string default_config = R"(#  _____     _ _
...
       port: 14000
-
+  
 
 # Monitoring configuration
 monitoring:
-  # Timeout for topic monitoring in ms (Default: 5000), increase in 1000er steps
-  timeout: 5000
+  # Timeout for topic monitoring in ms (Default: 1000), increase in 1000er steps
+  timeout: 1000
```
Comment (KerstinKeller) [inline ecal/core/src/config/default_config.h:50]: Didn't rex just change this back to 5000

**Review by KerstinKeller — COMMENTED** [review]
Will have to review again together .
I like the generation of the yam file, however it will be problematic during cross compilation.
We should consider to still have it as part of the repo (checked in)...

File: CMakeLists.txt [hunk 4]
```diff
@@ -103,7 +103,7 @@ option(ECAL_USE_CLOCKLOCK_MUTEX                "Use native mutex with monotonic
 set(ECAL_CORE_HAS_PROTOBUF                                                                                         ON)
 set(ECAL_CORE_BUILD_SAMPLES                                                                          ${BUILD_SAMPLES})
 set(ECAL_CORE_BUILD_TESTS                                                                         ${BUILD_ECAL_TESTS})
-set(ECAL_CORE_CONFIGURATION                                                                                        ON)
+option(ECAL_CORE_CONFIGURATION                                                                                     ON)
```
Comment (KerstinKeller) [inline CMakeLists.txt:106]: For our build, this should not be an option, we *always* what to build with the config -> yaml-cpp will be a fixed dependency for the core build.

File: ecal/core/include/ecal/config/configuration.h [hunk 5]
```diff
@@ -48,33 +48,29 @@ namespace eCAL
...
+    Service::Configuration        service;
+    Application::Configuration    application;
+    Logging::Configuration        logging;
+    Cli::Configuration            command_line_arguments;
 
     ECAL_API Configuration();
     ECAL_API Configuration(int argc_, char** argv_);
     ECAL_API Configuration(const std::vector<std::string>& args_);
 
     ECAL_API void InitConfigWithDefaultYaml();
```
Comment (KerstinKeller) [inline ecal/core/include/ecal/config/configuration.h:66]: Naming is a bit missleading. `InitFromConfig`, `InitConfigFromFile(const std::string& yaml_path_)` -> `InitFromFile`.

-> Also, do they still need to be member functions? Maybe free functions are better suited?

File: ecal/core/src/config/ecal_config_initializer.cpp [hunk 6]
```diff
@@ -201,52 +220,38 @@ namespace eCAL
       const std::string yaml_path = checkForValidConfigFilePath(yaml_path_);
       if (!yaml_path.empty())
       {
+#ifdef ECAL_CORE_CONFIGURATION
         eCAL::Config::YamlFileToConfig(yaml_path, *this);
         ecal_yaml_file_path = yaml_path;
+#else
+        std::cout << "Yaml file found at " << yaml_path << " but eCAL core configuration is not enabled." << "\n";
```
Comment (KerstinKeller) [inline ecal/core/src/config/ecal_config_initializer.cpp:227]: prefer eCAL log over cout?

File: ecal/core/src/ecal_def.h [hunk 7]
```diff
@@ -27,6 +27,7 @@
 #include <ecal/ecal_tlayer.h>
 #include <ecal/ecal_log_level.h>
 #include <ecal/ecal_os.h>
+#include <array>
```
Comment (KerstinKeller) [inline ecal/core/src/ecal_def.h:30]: not necessary

File: ecal/core/src/types/ecal_custom_data_types.cpp [hunk 8]
```diff
@@ -44,7 +45,7 @@ namespace eCAL
   {
 
     // IpAddressV4 definitions
-    IpAddressV4::IpAddressV4() : IpAddressV4(NET_UDP_MULTICAST_GROUP) {};
+    IpAddressV4::IpAddressV4() : IpAddressV4( "192.168.0.1" ) {};
```
Comment (KerstinKeller) [inline ecal/core/src/types/ecal_custom_data_types.cpp:48]: we should remove this constructor....

File: ecal/tests/cpp/config_test/CMakeLists.txt [hunk 9]
```diff
@@ -21,30 +21,34 @@ project(test_config)
 find_package(Threads REQUIRED)
 find_package(GTest REQUIRED)
 find_package(tclap REQUIRED)
-find_package(yaml-cpp REQUIRED)
+
+if(ECAL_CORE_CONFIGURATION)
+  find_package(yaml-cpp REQUIRED)
+endif()
 
 include(${ECAL_PROJECT_ROOT}/thirdparty/yaml-cpp/compatibility-yaml-cpp.cmake)
```
Comment (KerstinKeller) [inline ecal/tests/cpp/config_test/CMakeLists.txt:29]: this should be moved inside `if (ECAL_CORE_CONIFGURATION)` condition

**Review by KerstinKeller — COMMENTED** [review]
Let's address the changes...

File: ecal/core/include/ecal/config/configuration.h (hunk 5 above)
Comment (Peguen) [reply ecal/core/include/ecal/config/configuration.h:66]: As we want to store the loaded file path in a protected member, so that the user cannot modify it, we need it currently as member function.

File: ecal/core/src/ecal.cpp [hunk 10]
```diff
@@ -132,11 +132,6 @@ namespace eCAL
   **/
   int Initialize(eCAL::Configuration& config_, const char *unit_name_ /*= nullptr*/, unsigned int components_ /*= Init::Default*/)
   {
-    if (g_globals() == nullptr)
-    {
-      InitGlobals();
-    }
-
```
Comment (KerstinKeller) [inline ecal/core/src/ecal.cpp:139]: @Peguen why is this removed?

Comment (Peguen) [reply ecal/core/src/ecal.cpp:139]: I have put the InitGlobals() in the Configuration object. Reason was to use the eCAL logging mechanism even when eCAL::Initialize() was not called before.

**Review by KerstinKeller — COMMENTED** [review]
Can you create a test that will induce the crash?
Or at least later for validation?

**Comment by rex-schilasky** [comment]
> Can you create a test that will induce the crash? Or at least later for validation?

We should check this later in the common config validation function.

**Review by KerstinKeller — APPROVED** [review]
As discussed, we need to look at the initialization phase after.
